_posts/2020-09-08-communication-backend-comparison.md

```diff
@@ -0,0 +1,114 @@
+supports both one and multiple processes per node, in addition to multi-node training, communication between workers is crucial
+and will heavily affect performance, notably for compute bound training algorithms.
+
+This Blog post addresses and analyzes the raw performance of different communication backends, used to transmit tensors and other
+information between the workers.
+
+<!--more-->
+
+
+## Communication Backends
+
+Currently, MLBench supports 3 communication backends out of the box:
+
+* MPI, or Message Passing Interface (using [OpenMPI](https://www.open-mpi.org/) 's implementation)
+* GLOO, available with PyTorch
+* [NCCL](https://developer.nvidia.com/nccl), high-speed connectivity between GPUs if used with correct hardware
+
+Each backend presents its benefits and disadvantages, and is designed for specific
+use-cases, and those will be reflected in the results.
+
+###### Differences
+
+The following table illustrates the main differences between the 3 backends.
+
+| Backend | Comm. Functions | Optimized for | Float32 | Float16|
+|---------|-----------------|---------------|---------|--------|
+| MPI     |   All           |  CPU, GPU          | Yes     | No     |
+| GLOO    | All (on CPU), broadcast & all-reduce (on GPU)    |  CPU          | Yes     | Yes    |
+| NCCL    | broadcast, all reduce, reduce and all gather (on GPU) | GPU only | Yes | Yes
+
+As we can see, each has at least one main advantage, and must be used in specific cases.
+
+It is also important to note that PyTorch (built from source) comes with NCCL and GLOO pre-installed, so it can be
+more convenient for a user to use one of those two. Otherwise, MPI needs to be compiled and installed on the machine.
+
+---
+
+## Experiments
+
+In order to evaluate the performance of communication backends, we have created a dummy task called [Task 0](https://mlbench.readthedocs.io/en/latest/benchmark-tasks.html#task-0-communication-backend-raw-performance),
+which repeatedly sends random tensors of increasing sizes using an `all reduce` operation. This means that each worker shares his tensor with all other workers, and sums all the received tensors.
+
+The time taken for this operation is accurately measured 100 times for each sent tensor on each worker, and averaged to get a statistically significant estimation of communication times.
+
+To obtain those results, we have used the following hardware/software:
+
+- `PyTorch` deep learning framework
+- `n1-standard-4` (4 cores, 15GB RAM) machines on Google Cloud.
+- `NVIDIA® Tesla® T4` (16GB GDDR6, Turing arch)
+
+## Results
+
+As stated above, we compare the times of communication for different tensor types and backends.
+
+There are 4 tensor type: `Float16` & `Float32` CPU or GPU tensors.
+
+### CPU vs GPU tensors?
+
+Only MPI and GLOO support communication of CPU tensors, while NCCL requires the use of GPU acceleration. This is a great advantage, as CPU training is less costly
```

Comment:
say that the advantage is that MPI and GLOO support both kinds, vs NCCL just one